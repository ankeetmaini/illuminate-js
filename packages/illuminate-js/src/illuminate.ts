import { Token } from './Token';
import { add as addHook, run as runHook, reset as resetHooks } from './hooks';
import { encode } from './utils';

export interface TokenObject {
    pattern: RegExp;
    inside?: Definition;
    lookbehind?: boolean;
    greedy?: boolean;
    alias?: string;
}

export type Tokens = RegExp | TokenObject | Array<RegExp | TokenObject>;

export type Definition = Map<string, Tokens | Map<string, Tokens>>;

export type Plugin = (a: typeof addHook) => void;

export const languages = new Map<string, Definition>();

/**
 * Converts the code to an `Array<string | Token>`, which can be used to write a render.
 * For example, `react-illuminate` uses this to tokenize the code string and render it using react components.
 *
 * @param text    The code to be tokenized
 * @param grammar The definition to be used for tokenizing
 *
 * @private
 *
 * @example
 * import { tokenize } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * const code = `
 * ...
 * `;
 *
 * tokenize(code, javascript); // returns Array<string | Token>
 */
export function tokenize(text: string, grammar: Definition): Array<string | Token> {
    if (!(grammar instanceof Map)) {
        throw new Error('A grammar must be a Map');
    }

    const startArray: Array<string | Token> = [text];

    const rest = grammar.get('rest');

    if (rest instanceof Map) {
        for (const [token, def] of rest as Definition) {
            grammar.set(token, def);
        }

        //TODO: check if we need to shallow copy the grammar
        grammar.delete('rest');
    }

    tokenloop: for (const [token, def] of grammar) {
        const patterns = Array.isArray(def) ? def : [def];

        for (let j = 0; j < patterns.length; ++j) {
            const pattern = patterns[j];
            let lookbehindLength: number = 0;

            let regex = pattern instanceof RegExp ? pattern : (pattern as TokenObject).pattern;

            const { inside, lookbehind, alias, greedy } = <TokenObject>pattern;

            if (greedy && !regex.global) {
                const flags = (regex.toString().match(/[imuy]*$/) || [''])[0];

                regex = RegExp(regex.source, `${flags}g`);
            }

            for (let i = 0; i < startArray.length; i++) {
                // Donâ€™t cache length as it changes during the loop
                const node = startArray[i];

                // Something went terribly wrong, ABORT, ABORT!
                if (startArray.length > text.length) {
                    break tokenloop;
                }

                if (node instanceof Token) {
                    continue;
                }

                regex.lastIndex = 0;

                let match: RegExpExecArray | string | null = regex.exec(node);

                if (!match) {
                    continue;
                }

                if (lookbehind) {
                    lookbehindLength = match[1].length;
                }

                const stringFrom: number = match.index - 1 + lookbehindLength;

                match = match[0].slice(lookbehindLength);

                const stringTo = stringFrom + match.length;
                const before = node.slice(0, stringFrom + 1);
                const after = node.slice(stringTo + 1);

                const inserts = [];

                if (before) {
                    inserts.push(before);
                }

                const wrapped = new Token(token, inside ? tokenize(match, inside) : match, alias);

                inserts.push(wrapped);

                if (after) {
                    inserts.push(after);
                }

                startArray.splice(i, 1, ...inserts);
            }
        }
    }

    return startArray;
}

export interface HighlightEnv {
    code: string;
    highlightedCode: string;
    grammar: Definition;
    language: string;
}

/**
 * Highlights the given code. Returns the highlighted code in `string` format.
 *
 * @param text      The code string that is to be highlighted.
 * @param language  Name of the definition to be used. This name corresponds to
 *                  the name given to the definition using `addLanguage`.
 *
 * @example
 * import { addLanguage, highlight } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * addLanguage('javascript', javascript);
 * addLanguage('js', javascript);
 *
 * const code = `
 * ...
 * `;
 *
 * highlight(code, 'js');
 */
export function highlight(text: string, language: string): string {
    const grammar = languages.get(language);

    if (!grammar) {
        throw new Error(
            `Your are trying to use ${language} language but its definition could not be found. Please make sure to use "addLanguage" to add your definition`
        );
    }

    let env: HighlightEnv = {
        grammar,
        language,
        code: text,
        highlightedCode: ''
    };

    env = <HighlightEnv>runHook('before-highlight', env) || env;

    const tokens = tokenize(env.code, env.grammar); // Specific for plugins
    const highlightedCode = Token.stringify(encode(tokens), language);

    env = Object.assign({}, env, { highlightedCode: highlightedCode }) || env;
    env = <HighlightEnv>runHook('after-highlight', env) || env;

    return env.highlightedCode; // Specific for plugins
}

/**
 * Helper to add a plugin.
 *
 * @param plugin The plugin to be added.
 *               See [Plugins](./#/plugins/) for more details.
 *
 * @example
 * import { addPlugin } from 'illuminate-js';
 * import { showLanguage } from 'illuminate-js/lib/plugins/showLanguage';
 *
 * addPlugin(showLanguage);
 */
export function addPlugin(plugin: Plugin) {
    if (typeof plugin !== 'function') {
        throw new Error('Given Plugin must be a function');
    }

    plugin(addHook);
}

/**
 * Removes all the plugins.
 *
 * @example
 * import { resetPlugins } from 'illuminate-js';
 *
 * resetPlugins();
 */
export function resetPlugins() {
    resetHooks();
}

/**
 * Add a language definition to a illuminate.
 * You can add same definition multiple times with different names
 *
 * @param name  The name of the language being added.
 *              This name will be used by `highlight` function.
 * @param def   The language `Definition` to be added.
 *
 * @example
 * import { addLanguage } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * addLanguage('javascript', javascript);
 * addLanguage('js', javascript);
 */
export function addLanguage(name: string, def: Definition) {
    languages.set(name, def);
}

/**
 * Returns the language definition which was added using `addLanguage`.
 * Returns the `Definition` if found, else returns `undefined`.
 *
 * @param name The name of the language definition
 *
 * @private
 *
 * @example
 * import { addLanguage, getLanguage } from 'illuminate-js';
 * import { javascript } from 'illuminate-js/lib/languages';
 *
 * addLanguage('javascript', javascript);
 * addLanguage('js', javascript);
 *
 * getLanguage('js'); // returns javascript definition
 */
export function getLanguage(name: string): Definition | undefined {
    return languages.get(name);
}
