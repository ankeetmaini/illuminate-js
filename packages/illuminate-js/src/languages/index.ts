// This is an autogenerated file, please do not change anything here

import { apacheconf } from './apacheconf';
import { bash } from './bash';
import { clike } from './clike';
import { css } from './css';
import { diff } from './diff';
import { haskell } from './haskell';
import { http } from './http';
import { ini } from './ini';
import { javascript } from './javascript';
import { json } from './json';
import { jsx } from './jsx';
import { less } from './less';
import { makefile } from './makefile';
import { markdown } from './markdown';
import { markup } from './markup';
import { matlab } from './matlab';
import { php } from './php';
import { python } from './python';
import { ruby } from './ruby';
import { scss } from './scss';
import { sql } from './sql';
import { typescript } from './typescript';
import { yaml } from './yaml';

export {
    apacheconf,
    bash,
    clike,
    css,
    diff,
    haskell,
    http,
    ini,
    javascript,
    json,
    jsx,
    less,
    makefile,
    markdown,
    markup,
    matlab,
    php,
    python,
    ruby,
    scss,
    sql,
    typescript,
    yaml,
    markup as html,
    javascript as js,
    json as json5,
    markdown as md,
    typescript as ts
};
