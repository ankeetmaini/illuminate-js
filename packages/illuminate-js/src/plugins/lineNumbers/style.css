pre[class*='language-'] {
    position: relative;
}

.line-numbers-rows {
    position: absolute;
    pointer-events: none;
    top: 0;
    left: -3.8em;
    width: 3em; /* works for line-numbers below 1000 lines */
    letter-spacing: -1px;
    border-right: 1px solid #999;
    counter-reset: linenumber;
}

.line-numbers-rows > .line-number {
    pointer-events: none;
    display: block;
    counter-increment: linenumber;
}

.line-numbers-rows > .line-number::before {
    content: counter(linenumber);
    color: #999;
    display: block;
    padding-right: 0.8em;
    text-align: right;
}
