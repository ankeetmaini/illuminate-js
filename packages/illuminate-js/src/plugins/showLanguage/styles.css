pre[class*='language-'] {
    position: relative;
}
