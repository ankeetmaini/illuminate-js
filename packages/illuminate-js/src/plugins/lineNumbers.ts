import { Plugin, HighlightEnv } from '../illuminate';

export const lineNumber: Plugin = add => {
    add('after-highlight', env => {
        const match = (env as HighlightEnv).code.match(/\n(?!$)/g);
        const linesNum = match ? match.length + 1 : 1;
        let lines = '';

        for (let i = 0; i < linesNum; i++) {
            lines += `<span class="line-number"></span>`;
        }

        (env as HighlightEnv).highlightedCode += `<span class="line-number-rows">${lines}</span>`;
    });
};
