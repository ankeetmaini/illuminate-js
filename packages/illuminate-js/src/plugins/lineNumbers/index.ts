import { Plugin, HighlightEnv } from '../../illuminate';

export const lineNumbers: Plugin = add => {
    add('after-highlight', env => {
        const match = (env as HighlightEnv).code.match(/\n(?!$)/g);
        const linesNum = match ? match.length + 1 : 1;
        let lines = '';

        for (let i = 0; i < linesNum; i++) {
            lines += `<span class="line-number"></span>\n`;
        }

        (env as HighlightEnv).highlightedCode += `\n<span class="line-number-rows">\n${lines}</span>`;
    });
};
