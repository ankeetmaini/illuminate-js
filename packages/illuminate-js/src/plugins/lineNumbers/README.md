## Line Numbers Plugin

This plugin can be used to show line numbers with the highlighted code.

### Usage

1. Add the plugin

```js
import { addPlugin } from 'illuminate-js';
import { lineNumbers } from 'illuminate-js/lib/plugins/lineNumbers';

addPlugin(lineNumbers);
```

2. Include CSS file from `illuminate-js/lib/plugins/lineNumbers/style.css` or use it an reference to write your own styles.
