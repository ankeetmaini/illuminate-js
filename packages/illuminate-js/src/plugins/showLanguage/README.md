## Show Language Plugin

This plugin can be used to show the language of the highlighted code.

### Usage

1. Add the plugin

```js
import { addPlugin } from 'illuminate-js';
import { showLanguage } from 'illuminate-js/lib/plugins/showLangauage';

addPlugin(showLanguage);
```

2. Include CSS file from `illuminate-js/lib/plugins/showLangauage/style.css` or use it an reference to write your own styles.
