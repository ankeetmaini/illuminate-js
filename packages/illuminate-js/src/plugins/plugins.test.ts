import { addPlugin, resetPlugins, addLanguage, highlight } from '../illuminate';
import { json } from '../languages/json';

import { lineNumbers } from './lineNumbers';
import { showLanguage } from './showLanguage';

addLanguage('json', json);

const code = `[
  {
    "id": 1,
    "first_name": "Edan",
    "last_name": "Lauks",
    "email": "elauks0@zimbio.com"
  }, {
    "id": 2,
    "first_name": "Maggi",
    "last_name": "Osipenko",
    "email": "mosipenko1@harvard.edu"
  }
]`;

describe('Plugins Tests', () => {
    afterEach(() => {
        resetPlugins();
    });

    test('`lineNumbers` plugins shows line numbers', () => {
        addPlugin(lineNumbers);
        const hc = highlight(code, 'json');
        const match = hc.match(/<span class="line-number"><\/span>/g) || [];

        expect(match.length).toBe(13);
        expect(hc).toMatchSnapshot();
    });

    test('`showLanguage` plugins the highlighted language', () => {
        addPlugin(showLanguage);
        const hc = highlight(code, 'json');

        expect(hc.includes('<span class="show-language">json</span>')).toBe(true);
        expect(hc).toMatchSnapshot();
    });
});
