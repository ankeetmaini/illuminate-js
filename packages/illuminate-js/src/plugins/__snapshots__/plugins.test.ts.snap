// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugins Tests \`lineNumbers\` plugins shows line numbers 1`] = `
"<span class=\\"token punctuation brackets-square\\" >[</span>
  <span class=\\"token punctuation brackets-braces\\" >{</span>
    <span class=\\"token property\\" >\\"id\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token number\\" >1</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"first_name\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"Edan\\"</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"last_name\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"Lauks\\"</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"email\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"elauks0@zimbio.com\\"</span>
  <span class=\\"token punctuation brackets-braces\\" >}</span><span class=\\"token punctuation\\" >,</span> <span class=\\"token punctuation brackets-braces\\" >{</span>
    <span class=\\"token property\\" >\\"id\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token number\\" >2</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"first_name\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"Maggi\\"</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"last_name\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"Osipenko\\"</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"email\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"mosipenko1@harvard.edu\\"</span>
  <span class=\\"token punctuation brackets-braces\\" >}</span>
<span class=\\"token punctuation brackets-square\\" >]</span>
<span class=\\"line-number-rows\\">
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
<span class=\\"line-number\\"></span>
</span>"
`;

exports[`Plugins Tests \`showLanguage\` plugins the highlighted language 1`] = `
"<span class=\\"show-language\\">json</span><span class=\\"token punctuation brackets-square\\" >[</span>
  <span class=\\"token punctuation brackets-braces\\" >{</span>
    <span class=\\"token property\\" >\\"id\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token number\\" >1</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"first_name\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"Edan\\"</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"last_name\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"Lauks\\"</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"email\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"elauks0@zimbio.com\\"</span>
  <span class=\\"token punctuation brackets-braces\\" >}</span><span class=\\"token punctuation\\" >,</span> <span class=\\"token punctuation brackets-braces\\" >{</span>
    <span class=\\"token property\\" >\\"id\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token number\\" >2</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"first_name\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"Maggi\\"</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"last_name\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"Osipenko\\"</span><span class=\\"token punctuation\\" >,</span>
    <span class=\\"token property\\" >\\"email\\"</span><span class=\\"token operator\\" >:</span> <span class=\\"token string\\" >\\"mosipenko1@harvard.edu\\"</span>
  <span class=\\"token punctuation brackets-braces\\" >}</span>
<span class=\\"token punctuation brackets-square\\" >]</span>"
`;
